lib/src/arithmetic/leading_digit_anticipate.dart

```diff
@@ -0,0 +1,138 @@
+  Logic? get leadingOneB => output('leadingOneB');
+
+  /// If the reverse case is valid.
+  Logic? get validLeadOneB => output('validLeadOneB');
+
+  /// Input for telling LZA that a carry was seen on the add of inputs.
+  late final Logic? endAroundCarry;
+
+  /// Construct a leading-zero anticipate module that
+  /// predicts the number of leading zeros in the sum of
+  /// [a] and [b].
+  /// - if [endAroundCarry] is provided as input, then only the
+  /// [leadingOne] and [validLeadOne] outputs are set.  Otherwise,
+  /// the pairs of outputs for the [a] and [b] inputs are set. [endAroundCarry]
+  /// tells the LZA that a carry occurred when
+  /// subtracting [b] from [a], which means we know [a] was larger and
+  /// so we can simply pass out a computed [leadingOne] with its
+  /// corresponding [validLeadOne].
+  /// - Outputs [leadingOneA] which should be used if [a] > [b] (e.g., there
+  /// is a carry output from a ones-complement subtraction of [a] and [b].
+  /// - [leadingOneB] should be used if [b] >= [a].
+  LeadingZeroAnticipate(Logic aSign, Logic a, Logic bSign, Logic b,
+      {Logic? endAroundCarry, super.name = 'leading_zero_anticipate'}) {
+    aSign = addInput('aSign', aSign);
+    a = addInput('a', a, width: a.width);
+    bSign = addInput('bSign', bSign);
+    b = addInput('b', b, width: b.width);
+    this.endAroundCarry = (endAroundCarry != null)
+        ? addInput('endAroundCarry', endAroundCarry)
+        : null;
+
+    final aX = a.zeroExtend(a.width + 1).named('aX');
+    final bX = b.zeroExtend(b.width + 1).named('bX');
+    final t = (aX ^ mux(aSign ^ bSign, ~bX, bX)).named('t');
+    final zForward = (~aX & mux(aSign ^ bSign, bX, ~bX)).named('zerosForward');
+    final zReverse = ~bX & mux(aSign ^ bSign, aX, ~aX).named('zerosReverse');
+    final fForward = Logic(name: 'findForward', width: t.width);
+    final fReverse = Logic(name: 'findReverse', width: t.width);
+    fForward <= t ^ (~zForward << 1 | Const(1, width: t.width));
+    fReverse <= t ^ (~zReverse << 1 | Const(1, width: t.width));
+
+    final leadOneEncoderA = RecursiveModulePriorityEncoder(fForward.reversed,
+        generateValid: true, name: 'leadone_forward');
+    final leadingOneA = leadOneEncoderA.out;
+    final validLeadOneA = leadOneEncoderA.valid!;
+
+    final leadOneEncoderB = RecursiveModulePriorityEncoder(fReverse.reversed,
+        generateValid: true, name: 'leadone_reverse');
+    final leadingOneB = leadOneEncoderB.out;
+    final validLeadOneB = leadOneEncoderB.valid!;
+
+    if (this.endAroundCarry == null) {
+      addOutput('leadingOneA', width: log2Ceil(fForward.width + 1)) <=
+          leadingOneA;
+      addOutput('leadingOneB', width: log2Ceil(fReverse.width + 1)) <=
+          leadingOneB;
+      addOutput('validLeadOneA') <= validLeadOneA;
+      addOutput('validLeadOneB') <= validLeadOneB;
+    } else {
+      addOutput('leadingOne', width: log2Ceil(fForward.width + 1)) <=
```

Comment:
It seems like this "endAroundCarry" version could almost just be a wrapper around the non-endAroundCarry version of it?  Rather than a flag with optional outputs?

Reply:
I refactored into a base class that does a lot of the work, and then the two styles.  The challenge was that one of the classes produces MORE outputs with FEWER inputs, so its a bit complex to just have one extend the other.   So there was no clear way to order these classes.  So creating a 'Y' inheritance tree made more sense.